### PR TITLE
Allowing LSP clients without publish diagnostics data support to support code fixes

### LINKED ISSUE #2141: [feature] LSP fix support for LSP clients without diagnostic data support
# ⭐ Suggestion

I'm looking at the possibility of a VS extension which would use the LSP, unfortunately the fix rely on diagnostic data field which doesn't seem to be supported by Visual Studio's builtin LSP client

https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.diagnostic?view=visualstudiosdk-2022

# 💻 Use Cases

It's possible to use the LSP for diagnostics however the fixups don't exist.

### PR DESCRIPTION
**Pull request by reductor**
This should resolve #2141 which was a limitation for creating a Visual Studio extension for ast-grep using the builtin LSP client support, along with any LSP client which is prior to version 3.16.0 (14th Dec 2020)

I introduced some tests to support this however the existing tests and framework were a bit awkward for reusing data especially when things are interleaved with existing log messages and the did open ends up requiring "workspace/workspaceFolders" each time before it can get the published diagnostics (which might be a separate bug as it slows down getting diagnostics and make its it a multiple round trip request)

## Summary by CodeRabbit

- **New Features**
  - Per-document fix caching and improved code-action fallback so quick fixes are more reliable across clients.

- **Tests**
  - In-process LSP test harness with Content-Length framed JSON-RPC, codec round-trip check, and code-action application tests.

- **Chores**
  - Added development dependencies to support async testing and streaming utilities.

### CODE REVIEW

File: crates/lsp/src/lib.rs [hunk 1]
```diff
@@ -314,13 +328,33 @@ impl<L: LSPLang> Backend<L> {
...
     &self,
     uri: Uri,
     versioned: &mut VersionedAst<StrDoc<L>>,
   ) -> Option<()> {
     let diagnostics = self.get_diagnostics(&uri, versioned).unwrap_or_default();
     versioned.notes = self.build_notes(&diagnostics);
+    // Only save fixes if code action data is not supported
+    if !self.code_action_data_supported.load(std::sync::atomic::Ordering::Relaxed) {
+      versioned.fixes = self.build_fixes(&diagnostics);
+    }
```
[the remark this answers is not in the document]
Comment (reductor) [reply crates/lsp/src/lib.rs]: The data was still left, in case the support wasn't indicated but it would still support it transfering the diagnostic(s) verbatim

File: crates/lsp/src/lib.rs [hunk 2]
```diff
@@ -483,6 +519,14 @@ impl<L: LSPLang> Backend<L> {
...
     }
     let text_doc = params.text_document;
+
+    // If publish diagnostics do not support data, use cached fixes
+    let fixes_cache = if self.code_action_data_supported.load(std::sync::atomic::Ordering::Relaxed) {
+      None
+    } else {
+      Some(&self.map.get(text_doc.uri.as_str()).unwrap().fixes)
+    };
+
```
[the remark this answers is not in the document]
Comment (reductor) [reply crates/lsp/src/lib.rs]: The latest changes ensure the DashMap guard is kept alive.

**Review by HerringtonDarkholme — CHANGES_REQUESTED** [review]
* please fix linting and formatting errors
* I am not a fan of maintaining two sets of implementatiions. I would prefer just keeping the `fixCache` code path.

**Comment by reductor** [comment]
I've done some changes to remove the older approach and solely use the fixes cache, I plan on doing a bit more on adding further tests for the fixAll LSP action along with some other possibly edge cases now that the older method of using the data field is no longer used.

File: crates/lsp/tests/basic.rs [hunk 3]
```diff
@@ -357,3 +360,402 @@ async fn test_did_change_watched_files() {
...
+  );
+
+  for diagnostic in diagnostics.as_array().unwrap() {
+    let code_action = request_code_action(&mut client, file_uri, diagnostic).await;
+    let code_action = code_action.expect("No code action response");
+    assert_eq!(
+      code_action["result"].as_array().iter().len(),
+      1,
+      "Expected 1 code action per diagnostic"
+    );
```
[the remark this answers is not in the document]
Comment (reductor) [reply crates/lsp/tests/basic.rs]: This has been fixed in the latest changes.

**Comment by reductor** [comment]
> * please fix linting and formatting errors
> * I am not a fan of maintaining two sets of implementatiions. I would prefer just keeping the `fixCache` code path.

I've updated this again removing the data based implementation to now use just the fixCache and fixed the linting and formatting issues.

File: crates/lsp/tests/basic.rs [hunk 4]
```diff
@@ -357,3 +284,421 @@ async fn test_did_change_watched_files() {
...
+  for (start_line, start_char, _end_line, end_char, new_text) in all_edits {
+    let line = &lines[start_line];
+    let prefix = &line[..start_char];
+    let suffix = &line[end_char..];
+    lines[start_line] = format!("{}{}{}", prefix, new_text, suffix);
+  }
+
+  // Join lines back into a single string
+  lines.join("\n")
+}
```
[the remark this answers is not in the document]
Comment (HerringtonDarkholme) [reply crates/lsp/tests/basic.rs:436]: please create new issue to track this

File: crates/lsp/src/lib.rs [hunk 5]
```diff
@@ -425,24 +440,31 @@ impl<L: LSPLang> Backend<L> {
...
         }
-        let rewrite_data = RewriteData::from_value(d.data?)?;
         let fixed = rewrite_data.fixers.first()?.fixed.to_string();
-        let edit = TextEdit::new(d.range, fixed);
-        last = d.range.end;
+        let edit = TextEdit::new(*range, fixed);
+        last = range.end;
         Some(edit)
       })
       .collect();
```
[the remark this answers is not in the document]
Comment (reductor) [reply crates/lsp/src/lib.rs:470]: Stable iteration doesn't seem like it shouldn't be needed here, this sorting is primarily used for eliminating overlapping ranges with fix all

File: crates/lsp/src/utils.rs [hunk 6]
```diff
@@ -52,38 +48,51 @@ impl RewriteData {
...
+        .title
+        .unwrap_or_else(|| format!("Fix `{id}` with ast-grep"));
+      CodeAction {
+        title,
+        command: None,
+        diagnostics: None,
+        edit: Some(edit),
+        disabled: None,
+        kind: Some(CodeActionKind::QUICKFIX),
+        is_preferred: Some(i == 0), // mark the first fix as preferred
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/utils.rs:92]: good fix

File: crates/lsp/src/utils.rs [hunk 7]
```diff
@@ -52,38 +48,51 @@ impl RewriteData {
...
+        data: None,
+      }
+    });
   Some(actions.collect())
 }
```
[the remark this answers is not in the document]
Comment (HerringtonDarkholme) [reply crates/lsp/src/utils.rs:97]: no, keep implementation as is.

File: crates/lsp/tests/basic.rs [hunk 8]
```diff
@@ -357,3 +284,421 @@
...
+
+  fn encode(&mut self, item: serde_json::Value, dst: &mut BytesMut) -> Result<(), Self::Error> {
+    let json =
+      serde_json::to_string(&item).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+    let header = format!("Content-Length: {}\r\n\r\n", json.len());
+    dst.put(header.as_bytes());
+    dst.put(json.as_bytes());
+    Ok(())
+  }
+}
```
[the remark this answers is not in the document]
Comment (HerringtonDarkholme) [reply crates/lsp/tests/basic.rs:642]: it's only in test. it is okay

**Review by HerringtonDarkholme — APPROVED** [review]

**Comment by HerringtonDarkholme** [comment]
thanks
